packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java

```diff
@@ -142,6 +143,44 @@ public static boolean isRTL(MapBuffer attributedString) {
         == LayoutDirection.RTL;
   }
 
+  public static Layout.Alignment getTextAlignment(MapBuffer attributedString) {
+    boolean isRTL = isRTL(attributedString);
+    Layout.Alignment alignment = isRTL ? Layout.Alignment.ALIGN_OPPOSITE : Layout.Alignment.ALIGN_NORMAL;
+
+    MapBuffer fragments = attributedString.getMapBuffer(AS_KEY_FRAGMENTS);
+    if (fragments.getCount() != 0) {
+      MapBuffer fragment = fragments.getMapBuffer(0);
+      MapBuffer textAttributes = fragment.getMapBuffer(FR_KEY_TEXT_ATTRIBUTES);
+
+      if (textAttributes.contains(TextAttributeProps.TA_KEY_ALIGNMENT)) {
+        String alignmentAttr = textAttributes.getString(TextAttributeProps.TA_KEY_ALIGNMENT);
+
+        if (alignmentAttr.equals("center")) {
+          alignment = Layout.Alignment.ALIGN_CENTER;
+        } else if (alignmentAttr.equals("right")) {
+          alignment = isRTL ? Layout.Alignment.ALIGN_NORMAL : Layout.Alignment.ALIGN_OPPOSITE;
+        }
+      }
+    }
+
+    return alignment;
+  }
+
+  public static int getTextGravity(MapBuffer attributedString, int defaultValue) {
+    int gravity = defaultValue;
+    Layout.Alignment alignment = getTextAlignment(attributedString);
+
+    if (alignment == Layout.Alignment.ALIGN_NORMAL) {
+      gravity = Gravity.LEFT;
```

Comment:
Why `Gravity.LEFT` instead of `Gravity.START`?

Reply:
Note that only the position of the text itself is dependent on gravity, position of inline views is calculated in `measureText` method, which only relies on `Alignment`.

Reply:
This is consistent with the current behavior on the `main` branch, where `Layout.Alignment.ALIGN_NORMAL` is used all the time.

https://github.com/facebook/react-native/assets/21055725/bb3dda1d-a833-4d37-aee9-c387bfd2d5f5

Reply:
Looking at Android's code. I.e. internally, `NORMAL` should map to left or right depending on combination of paragraph alignment and layout direction. https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/text/Layout.java;l=628

I think we might be missing a call to set the layout direction on the layout builders. E.g. https://developer.android.com/reference/android/text/StaticLayout.Builder#setTextDirection(android.text.TextDirectionHeuristic)

Reply:
Changing that might have other implications. But it does clarify that text direction is currently always LTR by default, unless first script char is RTL.

I can't tell original intent of passing layout direction here. But it seems it is decoupled from text direction.

Reply:
I think the input layout direction roughly corresponds to "base direction" https://www.w3.org/International/articles/inline-bidi-markup/uba-basics#context

So we probably do want to be propagating this, but is is potentially a behavior change for how LTR scripts in RTL apps would previously work.

If we don't do that, we would need to resolve whether Android would choose LTR or RTL from script, instead of relying on passed layout direction.

Reply:
So, overall change I think would look like:
1. Set paragraph direction to `LTR` or `RTL` based on base direction. Strongly typed LTR or RTL characters will still display as RTL or LTR within that base direction though I think
2. Check how our actual span will be formatted, using something like TextDirectionHeuristic for maybe BidiFormatter.

Reply:
That change is probably worth being separate though. So we could avoid that here, but do the correct thing in RTL, by using the TextDirectionHeuristic to query what the text direction will end up as.

Reply:
> Changing that might have other implications. But it does clarify that paragraph direction is currently always LTR by default, unless first script char is RTL.

I've added the call to set text direction to `FIRSTSTRONG_LTR`, but it shouldn't have any implications, as it's the default anyway: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/text/StaticLayout.java;l=109;bpv=0;bpt=1

Reply:
> If we don't do that, we would need to resolve whether Android would choose LTR or RTL from script, instead of relying on passed layout direction.

<details>
<summary>Looks like the answer is that it does</summary>

```xml
<?xml version="1.0" encoding="utf-8"?>
<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    xmlns:tools="http://schemas.android.com/tools"
    android:id="@+id/main"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    tools:context=".MainActivity">

    <TextView
        android:id="@+id/text1"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:text="Hello World!"
        android:gravity="start"
        app:layout_constraintBottom_toTopOf="@id/text2"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintTop_toTopOf="parent" />

    <TextView
        android:id="@+id/text2"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:text="تخطيط الاختبار"
        android:gravity="start"
        app:layout_constraintBottom_toBottomOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintTop_toBottomOf="@id/text1" />

</androidx.constraintlayout.widget.ConstraintLayout>
```

<img width="430" alt="Screenshot 2024-04-24 at 15 25 52" src="https://github.com/facebook/react-native/assets/21055725/5c6c23e1-fbe8-475c-bd19-31b96a293db3">

</details>

Reply:
> That change is probably worth being separate though. So we could avoid that here, but do the correct thing in RTL, by using the TextDirectionHeuristic to query what the text direction will end up as.

Since it looks like Android will detect the gravity from the script, I've removed the custom RTL logic from my changes. The downside of this is that forcing RTL layout doesn't change the alignment of text anymore, but that's consistent with native Android app (the same screen as above, note that RTL is forced, as you can see on the status bar layout):

<img width="430" alt="Screenshot 2024-04-24 at 15 25 52" src="https://github.com/facebook/react-native/assets/21055725/24aaef0f-5c3e-407f-9c5b-9b8af7813fc0">

Is this ok, or should I try to make it align to the right when RTL is forced?